core/src/main/scala/kafka/log/Log.scala

```diff
@@ -1812,37 +1577,36 @@ class Log(@volatile private var _dir: File,
     endOffset: Long
   ): Unit = {
     logStartOffset = startOffset
-    nextOffsetMetadata = LogOffsetMetadata(endOffset, activeSegment.baseOffset, activeSegment.size)
-    recoveryPoint = math.min(recoveryPoint, endOffset)
+    localLog.updateLogEndOffset(endOffset)
```

Comment:
We need to preserve the LogOffsetMetadata for endOffset and use it to call updateHighWatermark.

Reply:
Sounds good. This can be updated to `updateHighWatermark(localLog.logEndOffsetMetadata)`.

Reply:
Done in 8ebb39f.